packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -2207,9 +2385,74 @@ function commitLayoutMountEffects_complete(
   root: FiberRoot,
   committedLanes: Lanes,
 ) {
+  const isModernRoot = (subtreeRoot.mode & ConcurrentMode) !== NoMode;
   while (nextEffect !== null) {
     const fiber = nextEffect;
-    if ((fiber.flags & LayoutMask) !== NoFlags) {
+
+    if (enableSuspenseLayoutEffectSemantics && isModernRoot) {
+      if (fiber.tag === OffscreenComponent) {
+        offscreenSubtreeWasHiddenStack.pop();
+        offscreenSubtreeIsHiddenStack.pop();
+        offscreenSubtreeWasHidden =
+          offscreenSubtreeWasHiddenStack.length > 0 &&
+          offscreenSubtreeWasHiddenStack[
+            offscreenSubtreeWasHiddenStack.length - 1
+          ];
+        offscreenSubtreeIsHidden =
+          offscreenSubtreeIsHiddenStack.length > 0 &&
+          offscreenSubtreeIsHiddenStack[
+            offscreenSubtreeIsHiddenStack.length - 1
+          ];
+      }
+    }
+
+    if (
+      enableSuspenseLayoutEffectSemantics &&
+      isModernRoot &&
+      offscreenSubtreeWasHidden &&
+      !offscreenSubtreeIsHidden
+    ) {
```

Comment:
I've verified in the built bundle that DCE works to correctly strip these conditional branches when `enableSuspenseLayoutEffectSemantics` is false.